File: tools/binding_compliance/conformance/families/log_collection.py
"""Log collection requires actual paths backed by complete move/copy inventories."""

from collections.abc import Mapping
from typing import Any

from ..coverage import CoveragePredicate, FamilyCoveragePolicy
from .file_operations import _files


def _observed(observation: Mapping[str, Any]) -> bool:
    """Reject invented discoveries and deletion or replacement of preserved source logs."""
    if set(observation) != {
        "before",
        "first",
        "afterFirst",
        "second",
        "files",
    } or not all(_files(observation[key]) for key in ("before", "afterFirst", "files")):
        return False
    before = {item["path"]: item["content"] for item in observation["before"]}
    first = {item["path"]: item["content"] for item in observation["afterFirst"]}
    after = {item["path"]: item["content"] for item in observation["files"]}
    for key, inventory in (("first", first), ("second", after)):
        paths = observation[key]
        if (
            not isinstance(paths, list)
            or paths != sorted(set(paths))
            or not all(path in inventory and path.endswith(".log") for path in paths)
        ):
            return False
    if not all(path in observation["second"] for path in observation["first"]):
        return False
    return all(
        (first.get(path) == content and after.get(path) == content)
        if path.startswith(("xse/", "custom/")) or not path.startswith("base/crash-")
        else (
            path not in first
            and first.get("base/Crash Logs/" + path.split("/")[-1]) == content
        )
        for path, content in before.items()
    )


_PYTHON_COLLECTOR = "parity:python:file_io.log_collection.PyLogCollector"
LOG_COLLECTION_COVERAGE_POLICY = FamilyCoveragePolicy(
    "log-collection",
    (
        CoveragePredicate(
            "targeted-resolver",
            "log-collection.collect",
            "log-collection.collect",
            "collection-effects",
            ("resolve_targeted_inputs",),
            _observed,
            runtime_operations=("resolve_targeted_inputs",),
        ),
        CoveragePredicate(
            "collected-logs",
            "log-collection.collect",
            "log-collection.collect",
            "collection-effects",
            (
                "LogCollector",
                "collect_all",
                "collect_crash_logs",
                "CRASH_AUTOSCAN_PATTERN",
                "CRASH_LOG_PATTERN",
            ),
            _observed,
            binding_obligation_ids=(
                "parity:cxx:1462d027a7cba8dd",
                "parity:cxx:acf327fd1cc5bdc2",
                "parity:cxx:8d5f7aa4b18d4e1f",
                "parity:node:scanlog.JsLogCollector",
                "parity:node:aux-phase4c-crash-autoscan-pattern",
                "parity:node:scanlog.patterns.CRASH_LOG_PATTERN",
                _PYTHON_COLLECTOR,
                *(
                    _PYTHON_COLLECTOR + "." + name
                    for name in (
                        "__init__",
                        "collect_all",
                        "collect_crash_logs",
                        "copy_from_xse_folder",
                        "crash_logs_dir",
                        "move_from_base_folder",
                        "pastebin_dir",
                    )
                ),
            ),
            runtime_operations=(
                None,
                "log_collector_new",
                "log_collector_collect_all",
                "log_collector_collect_crash_logs",
                *(
                    "PyLogCollector." + name
                    for name in (
                        "__init__",
                        "collect_all",
                        "collect_crash_logs",
                        "copy_from_xse_folder",
                        "crash_logs_dir",
                        "move_from_base_folder",
                        "pastebin_dir",
                    )
                ),
            ),
        ),
        CoveragePredicate(
            "configured-collector",
            "log-collection.configured",
            "log-collection.configured",
            "collection-effects",
            ("new_for_scan",),
            _observed,
            runtime_operations=("log_collector_new_for_scan",),
        ),
    ),
)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/log_collection.rs
//! Native log moves, preserved XSE source copies, and additive custom discovery.

use super::file_operations::{destination, files};
use super::{RunnerResult, invalid, text};
use classic_file_io_core::LogCollector;
use serde_json::{Value, json};
use std::{
    fs,
    path::{Path, PathBuf},
};

/// Materializes contained authored files before a native collection stage.
fn materialize(root: &Path, values: &Value) -> RunnerResult<()> {
    for (path, content) in values
        .as_object()
        .ok_or_else(|| invalid("collection files must be an object"))?
    {
        let target = destination(root, path)?;
        fs::create_dir_all(
            target
                .parent()
                .ok_or_else(|| invalid("file needs parent"))?,
        )?;
        fs::write(target, text(content)?.as_bytes())?;
    }
    Ok(())
}

/// Normalizes only actual returned paths while preserving the complete result set.
fn paths(root: &Path, values: Vec<PathBuf>) -> RunnerResult<Vec<String>> {
    let mut values = values
        .iter()
        .map(|path| {
            Ok(path
                .strip_prefix(root)?
                .to_string_lossy()
                .replace('\\', "/"))
        })
        .collect::<RunnerResult<Vec<_>>>()?;
    values.sort();
    Ok(values)
}

/// Runs full collection and individual copy/move/discovery APIs against two input stages.
pub(super) fn observe(fixture: &Value) -> RunnerResult<Value> {
    let temporary = tempfile::tempdir()?;
    let root = temporary.path();
    for folder in ["base", "xse", "custom"] {
        fs::create_dir(root.join(folder))?;
    }
    materialize(root, &fixture["files"])?;
    let yaml_root = tempfile::tempdir()?;
    // The explicit local-YAML XSE path avoids ambient installation discovery.
    // Keep bootstrap YAML outside the observed log tree because it contains an invocation path.
    let collector = if fixture["configured"] == true {
        fs::write(
            yaml_root.path().join("CLASSIC Fallout4 Local.yaml"),
            format!(
                "Game_Info:\n  Docs_Folder_XSE: {}\n",
                serde_json::to_string(&root.join("xse").to_string_lossy())?
            ),
        )?;
        LogCollector::new_for_scan(
            root.join("base"),
            yaml_root.path(),
            "Fallout4",
            "OG",
            None,
            Some(root.join("custom")),
        )
    } else {
        LogCollector::new(
            root.join("base"),
            Some(root.join("xse")),
            Some(root.join("custom")),
        )
    };
    let runtime = classic_shared_core::get_runtime();
    let mut result = json!({"before":files(root)?,"first":paths(root,runtime.block_on(collector.collect_all())?)?});
    if json!(paths(
        root,
        runtime.block_on(collector.collect_crash_logs())?
    )?) != result["first"]
    {
        return Err(invalid("crash discovery disagrees with full collection").into());
    }
    if collector.crash_logs_dir() != root.join("base/Crash Logs")
        || collector.pastebin_dir() != root.join("base/Crash Logs/Pastebin")
    {
        return Err(invalid("collector path accessor changed").into());
    }
    result["afterFirst"] = files(root)?;
    materialize(root, &fixture["later"])?;
    let moved = runtime.block_on(collector.move_from_base_folder())?;
    let copied = runtime.block_on(collector.copy_from_xse_folder())?;
    let later = fixture["later"]
        .as_object()
        .ok_or_else(|| invalid("later files must be an object"))?;
    if moved
        != later
            .keys()
            .filter(|path| path.starts_with("base/"))
            .count()
        || copied != later.keys().filter(|path| path.starts_with("xse/")).count()
    {
        return Err(invalid("individual move/copy counts disagree").into());
    }
    result["second"] = json!(paths(
        root,
        runtime.block_on(collector.collect_crash_logs())?
    )?);
    if json!(paths(root, runtime.block_on(collector.collect_all())?)?) != result["second"] {
        return Err(invalid("full discovery changed after individual operations").into());
    }
    result["files"] = files(root)?;
    let mut targeted = result["second"]
        .as_array()
        .ok_or_else(|| invalid("missing second collection"))?
        .iter()
        .map(|value| text(value).map(|path| root.join(path)))
        .collect::<RunnerResult<Vec<_>>>()?;
    targeted.push(root.join("missing-target.log"));
    let resolved =
        runtime.block_on(classic_file_io_core::log_collection::resolve_targeted_inputs(targeted));
    if json!(paths(root, resolved.logs)?) != result["second"]
        || resolved.rejected.len() != 1
        || resolved.rejected[0].path != root.join("missing-target.log")
        || resolved.rejected[0].reason != "path does not exist"
        || files(root)? != result["files"]
    {
        return Err(invalid("targeted resolver changed paths, rejection or owned files").into());
    }
    Ok(result)
}
